kv: remove outdated test

tbg opened the pull request:
the first call would signal the heartbeat goroutine to terminate,
which could recycle the txn metadata before the second call via
EndTransaction would happen (in that case, EndTransaction would
fail).

Switching the order removed that issue.

fixes #2541.

tamird commented:
This change looks like it changes the test to not test anything. 

> // TestTxnCoordIdempotentCleanup verifies that cleanupTxn is idempotent.

This change makes the test not test idempotence by ignoring the case that can return an error. Am I missing something?

tschottdorf commented:
no, the error was returned when the `EndTransaction` call would fail (which is unrelated to the test). All that matters for the test is that `cleanupTxn` gets called twice and it does: once after the `EndTransaction` completes successfully (when the heartbeat calls it) and once manually during the test.

bdarnell commented:
This test is strange; I don't think it makes any sense any more. When originally written (691a6a054), it worked because the transaction coordinator would recreate the transaction when `EndTransaction` was called after `txnMeta` was discarded. After transactions were pinned to a single coordinator (bbcfa1bd385), it *should* be failing and the only reason it's not is a race, in which the `EndTransaction` call runs before the heartbeat goroutine has shut down. Furthermore, since b38c5e557, I don't think we expect `cleanupTxn` to be called more than once, so it no longer needs to be idempotent. I think we should just get rid of this test.

tschottdorf commented:
You can still have multiple calls to `cleanupTxn`. The heartbeat loop and the main thread can both wind up calling it. I think we wrote the original test because the previous code in `cleanupTxn` could wind up double-closing a channel. So I don't think we should just drop the test.

bdarnell commented:
Not anymore: since b38c5e5577 only the main (`Send()`) thread calls `cleanupTxn`; the heartbeat thread never does.

tschottdorf commented:
Ah, right. Then I'll get rid of it and turn the double-call check in `cleanupTxn` into an assertion.

bdarnell commented on kv/txn_coord_sender.go:
```diff
@@ -501,7 +501,7 @@ func (tc *TxnCoordSender) cleanupTxn(trace *tracer.Trace, txn proto.Transaction)
 	txnMeta, ok := tc.txns[id]
 	// Only clean up once per transaction.
```
This comment and the one on the last line of the method are not quite accurate. We may want to remove the `txnEnd` nil check completely (although it does improve the error message in race conditions: `txnEnd` becomes nil during `cleanupTxn`, but `ok` doesn't become false until the heartbeat goroutine has stopped. It's a panic either way, but this panic is nicer than a double-closed channel).

bdarnell commented:
LGTM

tschottdorf commented:
apparently it **is** still called more than once in some cases, but only `TestSingleKey` catches it. Going to take a look.

tschottdorf commented:
ah, ok. `cleanupTxn` isn't called multiple times, but the heartbeat stopping means that the txn is unregistered. If that happens before `cleanupTxn`, it won't find `txnMeta` and (now) `panic`. I'll soften the assertion.

tbg committed:
double-calls to cleanupTxn

remove a test which verified that multiple calls to cleanupTxn
were allowed. Since then, the heartbeat goroutine does not
call cleanupTxn any more, so multiple calls don't happen.

fixes #2541.
